### ISSUE TITLE
There should be a separate TenantState value for while executing setup

### ISSUE THREAD

**Issue by DanielStolt** [OrchardCMS/Orchard#5650]
Current enum looks like this:

    public enum TenantState {
        Uninitialized,
        Running,
        Disabled,
        Invalid
    }

I propose we introduce an additional value `Initializing`.

The new tenant state should be set by `SetupService` at the beginning of setup, and either set back to `Uninitialized` if the operation failed, or set to `Running` after setup has completed if the operation succeeded.

While a tenant is in `Initializing` state it should not receive any requests. Any request made to a tenant in this state should be responded to with an HTTP 503 Service Unavailable, and a corresponding entity/message explaning "this site is currently being initialized; please try again later" or a similar message.

This will avoid a quite common problem of multiple users trying to access a tenant, some of which may be unaware that someone else might already be running setup on it. Currently the result of this race condition seems arbitrary. Users might see the setup screen or might see a half-cooked tenant or might see an error of some form.

The SetupService should of course also make sure the tenant is in the correct state `Uninitialized` before running setup on it, which will avoid multiple setups being invoked in parallel either by command line or by some other means.

**Comment by Piedone** [comment]
When is this common specifically? The case that you have an uninitialized site and multiple people try to access it and feel like running the setup (all within the timespan of the first person starting the setup but it not being finished yet, so maybe 2 minutes) seems very unlikely to me - I certainly haven't seen such an issue, not even with DotNest.

That said I don't see an issue with implementing this as something "nice to have" (though I'd probably call the state Initializing) but only if it doesn't overcomplicate or otherwise negatively impact things for the vast majority of cases when the tenant is Running.

**Comment by PaulDevenney** [comment]
we have this issue repeatedly, particularly within our non-production environments. We rebuild our sites from recipe on deployment to ensure a "known good" version. This stops random tweaks to (for example) enabled features from interfering with our tests. We also have automated test suites that run against these environments, which do not know to stop if the site is being rebuilt. This process forms part of our continuous deployment process and allows us to be sure that we can always deploy the site from recipe, no matter what changes have been made.   

As a side note, only small recipes take around two minutes, larger recipes with a lot of content can take significantly more so the window is larger.

I'd also call the state `Initializing` :)

**Comment by DanielStolt** [comment]
To elaborate a bit, this is not only to avoid concurrent setups, it's also an information problem. Present company excluded, I've worked in a number of highly geographically/organizationally distributed projects where a dev/test environment is often used by many teams/people who are unaware of each other. In many such situations it's simply not feasible to coordinate things such that everybody are always immediately aware when a tenant is recooked. Those who don't know what's going on will see that the site is misbehaving, and raise alerts, support tickets etc. which then have to be dealt with, etc etc.

It's a lot better if anybody hitting the site while it's initializing are made aware of what's going on. This is more than a "nice to have" IMO, but of course that depends on what kinds of projects you have experience with. Aside from that, this is a matter of correctness in Orchard. It should not be possible to get it to do the wrong thing.

And yes `Initializing` is a much better name, thank you. :)

**Comment by Piedone** [comment]
I agree that it shouldn't be possible to get things wrong in Orchard and I see that this can be an issue with distributed teams.

In your case however @PaulDevenney the automated tests IMO should definitely know about the site being re-installed: otherwise, even if there is a designed error for when the setup runs it'll be just a trial and error.

**Comment by nomiali5** [comment]
I am facing the same issue with message "This tenant is currently initializing. Please try again later.", Its run perfectly on my local machine. But Its failed to start on my server within virtual directory. Is any way to fix it. Should I need to deploy my local files on to server?
